tests/keygen.c

```diff
@@ -0,0 +1,244 @@
+		retval = -2;
+		goto err;
+	}
+	if (1 != EVP_DigestSignInit(mdctx, NULL, EVP_sha256(), NULL, priv)) {
+		display_openssl_errors(__LINE__);
+		retval = -3;
+		goto err;
+	}
+	if (1 != EVP_DigestSignUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors(__LINE__);
+		retval = -4;
+		goto err;
+	}
+	if (1 != EVP_DigestSignFinal(mdctx, NULL, &slen)) {
+		display_openssl_errors(__LINE__);
+		retval = -5;
+		goto err;
+	}
+	if (!(sig = OPENSSL_malloc(sizeof(unsigned char) * (slen)))) {
+		display_openssl_errors(__LINE__);
+		retval = -6;
+		goto err;
+	}
+	if (1 != EVP_DigestSignFinal(mdctx, sig, &slen)) {
+		display_openssl_errors(__LINE__);
+		retval = -7;
+		printf("Sign fail\n");
+		goto err;
+	}
+	printf("Sign success\n");
+
+	if (1 != EVP_DigestVerifyInit(mdctx, NULL, EVP_sha256(), NULL, pub)) {
+		display_openssl_errors(__LINE__);
+		retval = -8;
+		goto err;
+	}
+	if (1 != EVP_DigestVerifyUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors(__LINE__);
+		retval = -9;
+		goto err;
+	}
+	if (1 == EVP_DigestVerifyFinal(mdctx, sig, slen)) {
+		printf("Verify success\n");
+		retval = 0;
+		goto err;
+	} else {
+		display_openssl_errors(__LINE__);
+		printf("Verify fail\n");
+		retval = -10;
+		goto err;
+	}
+
+err:
+	if(sig) OPENSSL_free(sig);
+	if(mdctx) EVP_MD_CTX_destroy(mdctx);
+	return retval;
+}
+
+int main(int argc, char* argv[])
+{
+	int ret = EXIT_FAILURE, res;
+	ENGINE* engine = NULL;
+
+	if (argc < 5) {
+		printf("Too few arguments\n");
+		printf("%s [TOKEN1] [KEY-LABEL] [PIN] [CONF] [module]\n", argv[0]);
+		goto cleanup;
+	}
+	const char *token_label = argv[1], *key_label = argv[2],
+		*efile = argv[4], *module = argv[5];
+	char *key_pass = argv[3];
+
+	res = CONF_modules_load_file(efile, "engines", 0);
+	if (res <= 0) {
+		printf("cannot load %s\n", efile);
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+
+	ENGINE_add_conf_module();
+#if OPENSSL_VERSION_NUMBER>=0x10100000
+	OPENSSL_init_crypto(OPENSSL_INIT_ADD_ALL_CIPHERS \
+		| OPENSSL_INIT_ADD_ALL_DIGESTS \
+		| OPENSSL_INIT_LOAD_CONFIG, NULL);
+#else
+	OpenSSL_add_all_algorithms();
+	OpenSSL_add_all_digests();
+	ERR_load_crypto_strings();
+#endif
+	ERR_clear_error();
+
+	ENGINE_load_builtin_engines();
+	engine = ENGINE_by_id("pkcs11");
+	if (engine == NULL) {
+		printf("Could not get engine\n");
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+
+	if (!ENGINE_ctrl_cmd_string(engine, "PIN", key_pass, 0)) {
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+	if (!ENGINE_ctrl_cmd_string(engine, "DEBUG_LEVEL", "7", 0)) {
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+	if (module) {
+		if (!ENGINE_ctrl_cmd_string(engine, "MODULE_PATH", module, 0)) {
+			display_openssl_errors(__LINE__);
+			goto cleanup;
+		}
+	}
+	if (!ENGINE_init(engine)) {
+		printf("Could not initialize engine\n");
+		display_openssl_errors(__LINE__);
+		goto cleanup;
+	}
+
+	/*
+	 * EC key generation test
+	 */
+	PKCS11_EC_KGEN ec = {
```

Comment:
Declare structures and variables at the beginning of the code block, following the C89/C90 standard.